types/src/proof/definition.rs

```diff
@@ -470,6 +471,143 @@ impl From<AccumulatorConsistencyProof> for crate::proto::types::AccumulatorConsi
+/// left and `Y` and `Z` on the right.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct AccumulatorRangeProof<H> {
+    /// The siblings on the left of the path from root to the first leaf. Siblings near the root
+    /// are at the beginning of the vector.
+    left_siblings: Vec<HashValue>,
+
+    /// The sliblings on the right of the path from root to the last leaf. Siblings near the root
+    /// are at the beginning of the vector.
+    right_siblings: Vec<HashValue>,
+
+    phantom: PhantomData<H>,
+}
+
+impl<H> AccumulatorRangeProof<H>
+where
+    H: CryptoHasher,
+{
+    /// Constructs a new `AccumulatorRangeProof` using `left_siblings` and `right_siblings`.
+    pub fn new(left_siblings: Vec<HashValue>, right_siblings: Vec<HashValue>) -> Self {
+        Self {
+            left_siblings,
+            right_siblings,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Verifies the proof is correct. The verifier needs to have `expected_root_hash`, the index
+    /// of the first leaf and all of the leaves in possession.
+    pub fn verify(
+        &self,
+        expected_root_hash: HashValue,
+        first_leaf_index: Option<u64>,
+        leaf_hashes: &[HashValue],
+    ) -> Result<()> {
+        if first_leaf_index.is_none() {
+            ensure!(
+                leaf_hashes.is_empty(),
+                "first_leaf_index indicated empty list while leaf_hashes is not empty.",
+            );
+            ensure!(
+                self.left_siblings.is_empty() && self.right_siblings.is_empty(),
+                "No siblings are needed.",
+            );
+            return Ok(());
+        }
+
+        ensure!(
+            self.left_siblings.len() <= MAX_ACCUMULATOR_PROOF_DEPTH,
+            "Proof has more than {} ({}) left siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.left_siblings.len(),
+        );
+        ensure!(
+            self.right_siblings.len() <= MAX_ACCUMULATOR_PROOF_DEPTH,
+            "Proof has more than {} ({}) right siblings.",
+            MAX_ACCUMULATOR_PROOF_DEPTH,
+            self.right_siblings.len(),
+        );
+        ensure!(
+            !leaf_hashes.is_empty(),
+            "leaf_hashes is empty while first_leaf_index indicated non-empty list.",
+        );
+
+        let mut left_sibling_iter = self.left_siblings.iter().rev().peekable();
+        let mut right_sibling_iter = self.right_siblings.iter().rev().peekable();
+
+        let mut first_pos = Position::from_leaf_index(
+            first_leaf_index.expect("first_leaf_index should not be None."),
+        );
+        let mut current_hashes = leaf_hashes.to_vec();
+        let mut parent_hashes = vec![];
+
+        // Keep reducing the list of hashes by combining all the children pairs, until there is
+        // only one hash left.
+        while current_hashes.len() > 1
+            || left_sibling_iter.peek().is_some()
+            || right_sibling_iter.peek().is_some()
```

Comment:
Do you feel it's better to calculate the root level and loop by num of levels up (and assert siblings are fully consumed afterwards)? Probably it's stricter and more straightforward that way.

Reply:
I was following `AccumulatorProof::verify` which does it in a similar way...